src/executor/evaluate/mod.rs

```diff
@@ -563,57 +510,39 @@ async fn evaluate_function<'a, T: 'static + Debug>(
-                    }
-                },
+                Some(chars) => Ok(eval_to_str!(chars).chars().collect::<Vec<char>>()),
                 None => Ok(" ".chars().collect::<Vec<char>>()),
             };
-            match eval_to_str(expr).await? {
-                Nullable::Value(v) => {
-                    if name == "LTRIM" {
-                        Ok(Value::Str(v.trim_start_matches(&pattern?[..]).to_string()))
-                    } else {
-                        Ok(Value::Str(v.trim_end_matches(&pattern?[..]).to_string()))
-                    }
-                }
-                Nullable::Null => Ok(Value::Null),
+
+            let string = eval_to_str!(expr);
+            if name == "LTRIM" {
+                Ok(Value::Str(
+                    string.trim_start_matches(&pattern?[..]).to_string(),
+                ))
+            } else {
+                Ok(Value::Str(
+                    string.trim_end_matches(&pattern?[..]).to_string(),
+                ))
             }
             .map(Evaluated::from)
         }
-        Function::Reverse(expr) => {
-            match eval_to_str(expr).await? {
-                Nullable::Value(v) => Ok(Value::Str(v.chars().rev().collect::<String>())),
-                Nullable::Null => Ok(Value::Null),
-            }
-        }
-        .map(Evaluated::from),
+        Function::Reverse(expr) => Ok(eval_to_str!(expr).chars().rev().collect::<String>())
+            .map(Value::Str)
+            .map(Evaluated::from),
 
         Function::Substr { expr, start, count } => {
-            let string = match eval_to_str(expr).await? {
-                Nullable::Value(v) => v,
-                Nullable::Null => {
-                    return Ok(Evaluated::from(Value::Null));
-                }
-            };
-            let start = match eval_to_integer(start).await? {
-                Nullable::Value(v) => v - 1,
-                Nullable::Null => return Ok(Evaluated::from(Value::Null)),
+            let string = eval_to_str!(expr);
+            let start = eval_to_integer!(start) - 1;
+
+            let count = match count {
+                Some(v) => eval_to_integer!(v),
+                None => string.len() as i64,
             };
 
-            let end = match count {
-                Some(expr) => match eval_to_integer(expr).await? {
-                    Nullable::Value(count) => {
-                        if count < 0 {
-                            return Err(EvaluateError::NegativeSubstrLenNotAllowed.into());
-                        }
-                        min(max(start + count, 0) as usize, string.len())
-                    }
-                    Nullable::Null => return Ok(Evaluated::from(Value::Null)),
-                },
-                None => string.len(),
+            let end = if count < 0 {
+                return Err(EvaluateError::NegativeSubstrLenNotAllowed.into());
+            } else {
+                min(max(start + count, 0) as usize, string.len())
```

Comment:
How about this code?

```suggestion
            let end = (count >= 0)
                .then(|| min(max(start + count, 0) as usize, string.len()))
                .ok_or(EvaluateError::NegativeSubstrLenNotAllowed)?;
```

Reply:
It is not related to the current PR, but it was seen during the code change.